src/ray/core_worker/reference_count.h

```diff
@@ -172,6 +172,14 @@ class ReferenceCounter : public ReferenceCounterInterface,
       const absl::optional<NodeID> &pinned_at_raylet_id = absl::optional<NodeID>())
       LOCKS_EXCLUDED(mutex_);
 
+  /// Remove reference for an object that we own. The reference will only be
+  /// removed if the object's ref count is 0. This should only be used when
+  /// speculatively adding an owned reference that may need to be rolled back, e.g. if
+  /// the creation of the corresponding Plasma object fails.
```

Comment:
Can you specify that otherwise all references should get cleaned up through ref counting?